File: laicode/q139/recursion.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        // assume root is not NULL
        int res = INT_MIN;
        maxPathSum(root, res);
        return res;
    }
private:
    int maxPathSum(TreeNode* root, int& res) {
        if (root == NULL) {
            return 0;
        }
        int l = maxPathSum(root->left, res);
        int r = maxPathSum(root->right, res);
        l = l > 0 ? l : 0;
        r = r > 0 ? r : 0;
        res = max(l + r + root->value, res);
        return max(l, r) + root->value;
    }
};

class TreeNode {
public:
    int value;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : value(v), left(NULL), right(NULL) {} 
};

int main() {

}

File: laicode/q139/Recursion.java
package laicode.q139;

public class Recursion {

    public int maxPathSum(TreeNode root) {
        // assume root is not null
        int[] max = {Integer.MIN_VALUE};
        maxPathSum(root, max);
        return max[0];
    }

    private int maxPathSum(TreeNode root, int[] max) {
        if (root == null) {
            return 0;
        }
        int left = maxPathSum(root.left, max);
        int right = maxPathSum(root.right, max);
        left = left > 0 ? left : 0;
        right = right > 0 ? right : 0;
        max[0] = Math.max(max[0], left + right + root.key);
        return Math.max(left, right) + root.key;
    }
}

class TreeNode {
    int key;
    TreeNode left;
    TreeNode right;

    TreeNode(int key) {
        this.key = key;
        left = right = null;
    }
}